### PR TITLE
csp-header@2

### PR DESCRIPTION
**Pull request by frux**
## What's happened?
I have rewritten `csp-header` in TypeScript.

#### Dropped support of `extend`
`extend` was marked as deprecated in previous versions. It doesn't work anymore. Use `presets` instead.
#### Dropped support of specifying presets as a string
`csp-header` used to require preset if you specify it as a string.
Before:
```js
{
    //...
    presets: ['csp-preset-myservice']
}
```
Now:
```js
{
    //...
    presets: [require('csp-preset-myservice')]
}
```
#### Calling with no arguments returns empty string

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,24 +53,6 @@ or as a keyed object:
...
-```js
-{
-    presets: {
-        superPuperService: 'super-puper-service' // takes node_modules/csp-preset-super-puper-service
-    }
-}
-```
-
 ## Preset format
 If you have a web-service feel free to publish preset of rules for using your service. For example your service is ``my-super-service.com``. Just publish preset ``csp-preset-my-super-service`` containing following code:
```
Comment (f0rmat1k) [inline README.md]: For example`,`

File: README.md [hunk 2]
```diff
@@ -81,18 +62,33 @@ modules.exports = {
...
-        'connect-src': ['test.com']
-    }
-});
+### Minimal supported version of node is 8
+
+### Dropped support of `extend`
+`extend` was marked as deprecated in previous versions. It doesn't work anymore. Use `presets` instead.
+
+### Dropped support of specifying presets as a string
+`csp-header` used to require preset if you specify it as a string. Now, please require it by yourself.
```
Comment (f0rmat1k) [inline README.md]: `please` is a little bit strange, `you should require` is better.

File: README.md [hunk 3]
```diff
@@ -81,18 +62,33 @@ modules.exports = {
...
+```
+Now:
+```js
+{
+    //...
+    presets: [require('csp-preset-myservice')]
+}
 ```
+
+### Calling with no arguments returns empty string
```
Comment (f0rmat1k) [inline README.md]: `an empty`? But i'm not sure

File: README.md [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # csp-header
-Content-Security-Policy header generator for Node.JS
+Content-Security-Policy header generator for Node
```
Comment (f0rmat1k) [inline README.md]: `Node.js` is best in my opinion

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	presets?: CSPPreset,
+	reportUri?: string
+}
+
+export type CSPDirectives = Partial<(
+	Record<CSPListDirectiveName, CSPListDirectiveValue> |
+	Record<CSPStringDirectiveName, CSPStringDirectiveValue> |
+	Record<CSPBooleanDirectiveName, CSPBooleanDirectiveValue>
+)>;
+export type CSPDirectiveName = keyof typeof ALLOWED_DIRECTIVES;
```
Comment (f0rmat1k) [inline src/index.ts]: Not important, but there is good practice to place types in separated file (turbo argument)

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+			...directiveValue1 as CSPListDirectiveValue,
+			...directiveValue2 as CSPListDirectiveValue
+		]);
+	}
+
+	return directiveValue2;
+}
+
+function getUniqRules(rules: CSPListDirectiveValue): CSPListDirectiveValue {
+	return Array.from(new Set(rules));
```
Comment (f0rmat1k) [inline src/index.ts:148]: I suppose garbage collector will work fine here, but maybe WeakSet for more sure?

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,62 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
```
Comment (f0rmat1k) [inline tsconfig.json]: https://github.com/microsoft/TypeScript-wiki/pull/182/files#diff-d9bc6bd0bb1889c1f902274da281a58eR22

"target": "es2017"

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "lib": ["es2018"],                             /* Specify library files to be included in the compilation. */
+    // "allowJs": true,                       /* Allow javascript files to be compiled. */
+    // "checkJs": true,                       /* Report errors in .js files. */
+    "jsx": "react",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
```
Comment (f0rmat1k) [inline tsconfig.json]: unnecessary rule

File: tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,62 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "lib": ["es2018"],                             /* Specify library files to be included in the compilation. */
+    // "allowJs": true,                       /* Allow javascript files to be compiled. */
```
Comment (f0rmat1k) [inline tsconfig.json]: Maybe should remove commented rules?

**Review by f0rmat1k — CHANGES_REQUESTED** [review]

File: tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,62 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "lib": ["es2018"],                             /* Specify library files to be included in the compilation. */
```
Comment (f0rmat1k) [inline tsconfig.json]: and they recomend lib es2017 https://github.com/microsoft/TypeScript-wiki/pull/182/files#diff-d9bc6bd0bb1889c1f902274da281a58eR27

File: src/index.ts (hunk 6 above)
Comment (frux) [reply src/index.ts:148]: It would work if rules was an array of objects. But it's an array of string values.

File: src/index.ts (hunk 5 above)
Comment (frux) [reply src/index.ts]: Nice argument!

**Comment by frux** [comment]
@f0rmat1k Thank you for your review! It was really useful.

**Pull request merged**